doc/code/qml_hf.rst

```diff
@@ -1,7 +1,160 @@
-qml.hf
-======
-
-.. currentmodule:: pennylane.hf
-
-.. automodapi:: pennylane.hf
-
+qml.hf
+======
+
+This module provides the functionality to perform differentiable Hartree-Fock (HF) calculations and
+construct molecular Hamiltonians that can be differentiated with respect to nuclear coordinates and
+basis set parameters.
+
+The differentiable aspect of the HF solver allows computing exact gradients with respect to
+molecular geometry, basis set, and circuit parameters simultaneously using the techniques of
+automatic differentiation available in `Autograd <https://github.com/HIPS/autograd>`__. This makes
+this HF solver more versatile and robust compared to non-differentiable tools which mainly rely
+on numerical approaches for computing gradients, with an extra computational cost. In particular,
+optimizing the basis set parameters available with the Hf solver allows reaching lower ground state
+energies without increasing the size of the basis set which leads to a larger number of number of
+qubits.
+
+.. currentmodule:: pennylane.hf
+
+.. automodapi:: pennylane.hf
+
+Using the differentiable HF solver
+----------------------------------
+
+The HF solver computes the integrals over basis functions, constructs the relevant matrices, and
+performs self-consistent-field iterations to obtain a set of optimized molecular orbital
+coefficients. These coefficients and the computed integrals over basis functions are used to
+construct the one- and two-body electron integrals in the molecular orbital basis, which can be
+used to generate a differentiable second-quantized Hamiltonian in the fermionic and qubit basis.
+
+The following code shows the construction of the Hamiltonian for the hydrogen molecule where the
+geometry of the molecule and the basis set parameters are all differentiable.
+
+.. code-block:: python3
+
+    import pennylane as qml
+    from pennylane import numpy as np
+
+    symbols = ["H", "H"]
+    geometry = np.array([[0.0, 0.0, 0.0], [0.0, 0.0, 2.0]], requires_grad=True)
+
+    # alpha and coeff are the exponents and contraction coefficients of the Gaussian functions
+    alpha = np.array([[3.42525091, 0.62391373, 0.1688554],
+                      [3.42525091, 0.62391373, 0.1688554]], requires_grad = True)
+    coeff = np.array([[0.15432897, 0.53532814, 0.44463454],
+                      [0.15432897, 0.53532814, 0.44463454]], requires_grad = True)
+
+We create a molecule object with differentiable atomic coordinates and basis set parameters and then
+construct the Hamiltonian.
+
+.. code-block:: python3
+
+    mol = qml.hf.Molecule(symbols, geometry, alpha=alpha, coeff=coeff)
+    args_mol = [geometry, alpha, coeff] # initial values of the differentiable parameters
+
+    hamiltonian = qml.hf.generate_hamiltonian(mol)(*args_mol)
+
+>>> print(hamiltonian)
+  ((-0.3596823592263728+0j)) [I0]
```

Comment:
Why are there double brackets here? I find that odd

Reply:
It is happening because the coefficients have a complex component, see for example:
```
c = np.array([0.2+0.j, -0.543+0.j])
o = [qml.PauliX(0), qml.PauliZ(0)]
H = qml.Hamiltonian(coeffs, obs)
print(H)
```
which gives
```
  ((-0.543+0j)) [Z0]
+ ((0.2+0j)) [X0]
```

It is easy to convert the coefficients to real and reconstruct the Hamiltonian. Will add it to our todo list if seems important.

Reply:
No, it's fine. I was just curious